File: apps/mobile/android/app/src/main/kotlin/open/fpvlabs/stera/video_picker/VideoPickerMethodChannelHandler.kt
package open.fpvlabs.stera.video_picker

import android.app.Activity
import android.content.Intent
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/**
 * Handles Flutter MethodChannel calls for VideoPicker operations.
 */
class VideoPickerMethodChannelHandler(
    activity: Activity,
    binaryMessenger: BinaryMessenger
) {
    companion object {
        private const val CHANNEL_NAME = "custom_document_picker"
    }

    private val videoPicker: VideoPicker = VideoPickerImpl(activity)
    private val methodChannel: MethodChannel = MethodChannel(binaryMessenger, CHANNEL_NAME)
    private var pendingResult: MethodChannel.Result? = null

    init {
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "pickVideos" -> handlePickVideos(result)
                "pickSingleVideo" -> handlePickSingleVideo(result)
                else -> result.notImplemented()
            }
        }
    }

    private fun handlePickVideos(result: MethodChannel.Result) {
        if (pendingResult != null) {
            result.error("ALREADY_ACTIVE", "A picker is already active", null)
            return
        }
        pendingResult = result
        videoPicker.pickVideos()
    }

    private fun handlePickSingleVideo(result: MethodChannel.Result) {
        if (pendingResult != null) {
            result.error("ALREADY_ACTIVE", "A picker is already active", null)
            return
        }
        pendingResult = result
        videoPicker.pickSingleVideo()
    }

    /**
     * Must be called from Activity.onActivityResult to handle picker results.
     */
    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        val (handled, result) = videoPicker.handleActivityResult(requestCode, resultCode, data)
        if (handled) {
            pendingResult?.success(result)
            pendingResult = null
        }
    }
}

File: apps/mobile/ios/Runner/video_picker/VideoPickerMethodChannelHandler.swift
import Foundation
import Flutter
import UIKit

/// Handles Flutter MethodChannel calls for VideoPicker operations.
/// Equivalent to Android's VideoPickerMethodChannelHandler.
class VideoPickerMethodChannelHandler: NSObject {
    
    private static let channelName = "custom_document_picker"
    
    private var videoPicker: VideoPicker
    private let methodChannel: FlutterMethodChannel
    private var pendingResult: FlutterResult?
    private var isPickingMultiple: Bool = false
    
    init(viewController: UIViewController, binaryMessenger: FlutterBinaryMessenger) {
        self.videoPicker = VideoPickerImpl(viewController: viewController)
        self.methodChannel = FlutterMethodChannel(
            name: VideoPickerMethodChannelHandler.channelName,
            binaryMessenger: binaryMessenger
        )
        
        super.init()
        
        // Set self as delegate
        videoPicker.delegate = self
        
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handleMethodCall(call: call, result: result)
        }
    }
    
    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "pickVideos":
            handlePickVideos(result: result)
        case "pickSingleVideo":
            handlePickSingleVideo(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Method Handlers
    
    private func handlePickVideos(result: @escaping FlutterResult) {
        if pendingResult != nil {
            result(FlutterError(code: "ALREADY_ACTIVE", message: "A picker is already active", details: nil))
            return
        }
        
        pendingResult = result
        isPickingMultiple = true
        videoPicker.pickVideos()
    }
    
    private func handlePickSingleVideo(result: @escaping FlutterResult) {
        if pendingResult != nil {
            result(FlutterError(code: "ALREADY_ACTIVE", message: "A picker is already active", details: nil))
            return
        }
        
        pendingResult = result
        isPickingMultiple = false
        videoPicker.pickSingleVideo()
    }
}

// MARK: - VideoPickerDelegate

extension VideoPickerMethodChannelHandler: VideoPickerDelegate {
    
    func didPickVideos(_ urls: [String]?) {
        pendingResult?(urls)
        pendingResult = nil
    }
    
    func didPickSingleVideo(_ url: String?) {
        pendingResult?(url)
        pendingResult = nil
    }
}
